crates/whiskers/src/grid/mod.rs

```diff
@@ -0,0 +1,181 @@
+///     [10.0, 10.0],
+///     Point::new(0.0, 0.0),
+/// );
+/// grid.init(None);
+/// ```
+pub struct Grid<'a, T> {
+    sketch: &'a mut Sketch,
+    columns: usize,
+    rows: usize,
+    size: GridSize,
+    gutter: [f64; 2],
+    position: Point,
+    /// List of grid cells filled with data of given type
+    pub data: Vec<GridCell<T>>,
+}
+
+/// Function to populate the grid with data of given type
+pub type GridInitFn<T> = fn(column: usize, row: usize, data: &Vec<GridCell<T>>) -> Option<T>;
+
+impl<'a, T> Grid<'a, T> {
+    /// Creates new instance of the grid with empty data vector
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        sketch: &'a mut Sketch,
+        columns: usize,
+        rows: usize,
+        size: GridSize,
+        gutter: [f64; 2],
+        position: Point,
+    ) -> Self {
+        Self {
+            sketch,
+            columns,
+            rows,
+            size,
+            gutter,
+            position,
+            data: vec![],
+        }
+    }
+
+    /// Fills data vector with grid cells. You can pass option with grid cell
+    /// filling function to pass data of your type to every grid cell
+    #[allow(clippy::cast_possible_truncation, clippy::cast_precision_loss)]
+    pub fn init(&mut self, init_fn: Option<GridInitFn<T>>) -> &mut Self {
+        let [module_width, module_height] = self.module_size();
+        let [gutter_width, gutter_height] = self.gutter;
+        let mut data: Vec<GridCell<T>> = vec![];
+
+        for row in 0..self.rows {
+            for column in 0..self.columns {
+                let pos_x = self.position.x()
+                    + (column as f64 * module_width + column as f64 * gutter_width);
+                let pos_y =
+                    self.position.y() + (row as f64 * module_height + row as f64 * gutter_height);
+
+                data.push(GridCell {
+                    column,
+                    row,
+                    canvas_position: Point::new(pos_x, pos_y),
+                    data: match init_fn {
+                        Some(f) => f(column, row, &data),
+                        None => None,
+                    },
+                    size: self.module_size(),
+                });
+            }
+        }
+
+        self.data = data;
+        self
+    }
+
+    /// Returns optional reference to a grid cell at specific column and row index
+    pub fn at(&mut self, column: usize, row: usize) -> Option<&mut GridCell<T>> {
+        self.data
+            .iter_mut()
+            .find(|cell| cell.column == column && cell.row == row)
+    }
+
+    /// Resets grid cell's data at specific column and row if applicable
+    pub fn reset_at(&mut self, column: usize, row: usize) {
+        let cell = self.at(column, row);
+
+        if let Some(cell) = cell {
+            cell.reset_data();
+        }
+    }
+
+    /// Resets all grid cell's data
+    pub fn reset(&mut self) {
+        self.data.iter_mut().for_each(GridCell::reset_data);
+    }
```

Comment:
Being a nice "container of user data" is tough and, again, best avoided altogether.